File: backend/app/embedder.py
import logging
import os
from typing import List, Optional
import httpx
from tenacity import retry, stop_after_attempt, wait_exponential

logger = logging.getLogger(__name__)

class BioBERTEmbedder:
    """Generates embeddings using BioBERT model via RunPod API."""
    
    def __init__(self):
        """Initialize the RunPod embedder client."""
        logger.info("Initializing RunPod embedder client")
        self.embedding_url = os.getenv("RUNPOD_EMBEDDING_URL")
        self.api_key = os.getenv("RUNPOD_EMBEDDING_KEY")
        
        if not self.embedding_url or not self.api_key:
            raise ValueError("RUNPOD_EMBEDDING_URL and RUNPOD_EMBEDDING_KEY must be set")
        
        # Initialize HTTP client with default headers
        self.client = httpx.Client(
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            }
        )
    
    def is_loaded(self) -> bool:
        """Check if the RunPod service is configured and accessible."""
        try:
            # Make a test request to the API
            response = self.client.get(f"{self.embedding_url}/health")
            return response.status_code == 200
        except Exception as e:
            logger.error(f"Error checking RunPod service: {str(e)}")
            return False
    
    @retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=4, max=10))
    def embed_text(self, text: str, max_length: int = 512) -> List[float]:
        """
        Generate embedding for the given text using RunPod API.
        
        Args:
            text: The text to embed
            max_length: Maximum token length (handled by RunPod service)
            
        Returns:
            Vector embedding as a list of floats
        """
        try:
            # Prepare request payload
            payload = {
                "text": text,
                "max_length": max_length
            }
            
            # Make request to RunPod API
            response = self.client.post(
                f"{self.embedding_url}/embed",
                json=payload,
                timeout=30.0  # 30 second timeout
            )
            
            # Check for successful response
            response.raise_for_status()
            
            # Parse response
            result = response.json()
            if "error" in result:
                raise RuntimeError(f"RunPod API error: {result['error']}")
            
            embeddings = result.get("embedding")
            if not embeddings:
                raise ValueError("No embedding returned from RunPod API")
            
            logger.info(f"Generated embedding with {len(embeddings)} dimensions via RunPod")
            return embeddings
            
        except Exception as e:
            logger.error(f"Error generating embedding via RunPod: {str(e)}")
            raise
    
    def embed_batch(self, texts: List[str], batch_size: int = 8) -> List[List[float]]:
        """
        Generate embeddings for a batch of texts.
        
        Args:
            texts: List of texts to embed
            batch_size: Number of texts to process at once
            
        Returns:
            List of vector embeddings
        """
        try:
            # Prepare batch payload
            payload = {
                "texts": texts,
                "batch_size": batch_size
            }
            
            # Make batch request to RunPod API
            response = self.client.post(
                f"{self.embedding_url}/embed_batch",
                json=payload,
                timeout=60.0  # 60 second timeout for batch processing
            )
            
            # Check for successful response
            response.raise_for_status()
            
            # Parse response
            result = response.json()
            if "error" in result:
                raise RuntimeError(f"RunPod API error: {result['error']}")
            
            embeddings = result.get("embeddings")
            if not embeddings:
                raise ValueError("No embeddings returned from RunPod API")
            
            logger.info(f"Generated {len(embeddings)} embeddings via RunPod")
            return embeddings
            
        except Exception as e:
            logger.error(f"Error generating batch embeddings via RunPod: {str(e)}")
            # Fall back to sequential processing if batch fails
            results = []
            for i in range(0, len(texts), batch_size):
                batch = texts[i:i+batch_size]
                batch_results = [self.embed_text(text) for text in batch]
                results.extend(batch_results)
            return results

File: backend/lib/embedder.js
import axios from 'axios';

export class EmbeddingService {
  constructor() {
    this.runpodUrl = process.env.RUNPOD_EMBEDDING_URL;
    this.runpodKey = process.env.RUNPOD_EMBEDDING_KEY;
    this.openaiKey = process.env.OPENAI_API_KEY;
    
    console.log('EmbeddingService initialized');
    console.log(`RunPod configured: ${!!this.runpodUrl}`);
    console.log(`OpenAI configured: ${!!this.openaiKey}`);
  }

  isConfigured() {
    return !!(this.runpodUrl && this.runpodKey) || !!this.openaiKey;
  }

  // CRITICAL FIX: Updated to accept userJWT parameter
  async generateEmbedding(text, userJWT = null) {
    if (!text || typeof text !== 'string') {
      throw new Error('Text input is required for embedding generation');
    }

    // Clean and truncate text if needed
    const cleanText = text.trim().substring(0, 8000); // Limit to prevent token overflow
    
    try {
      // CRITICAL FIX: Try TxAgent BioBERT first with user's JWT (medical-specific)
      if (this.runpodUrl && userJWT) {
        console.log('🔧 Using TxAgent BioBERT with user JWT authentication');
        return await this.generateTxAgentEmbedding(cleanText, userJWT);
      }
      
      // Fallback to OpenAI if no JWT or TxAgent not available
      if (this.openaiKey) {
        console.log('🔧 Falling back to OpenAI embedding service');
        return await this.generateOpenAIEmbedding(cleanText);
      }
      
      throw new Error('No embedding service configured or JWT token missing');
      
    } catch (error) {
      console.error('Embedding generation failed:', error);
      
      // If TxAgent fails, try OpenAI as fallback
      if (this.runpodUrl && this.openaiKey && userJWT) {
        console.log('TxAgent failed, trying OpenAI fallback...');
        try {
          return await this.generateOpenAIEmbedding(cleanText);
        } catch (fallbackError) {
          console.error('OpenAI fallback also failed:', fallbackError);
          throw new Error('All embedding services failed');
        }
      }
      
      throw error;
    }
  }

  // CRITICAL FIX: New method for TxAgent container with proper JWT authentication
  async generateTxAgentEmbedding(text, userJWT) {
    try {
      console.log('🚀 Calling TxAgent container for BioBERT embedding');
      console.log(`📍 TxAgent URL: ${this.runpodUrl}`);
      console.log(`🔐 Using user JWT authentication: ${userJWT ? 'YES' : 'NO'}`);
      console.log(`📝 Text length: ${text.length} characters`);

      // CRITICAL FIX: Use user's Supabase JWT instead of RunPod API key
      const response = await axios.post(
        `${this.runpodUrl}/embed`,
        {
          file_path: `inline_text_${Date.now()}`,
          metadata: {
            text_length: text.length,
            source: 'direct_embedding',
            inline_text: text,
            timestamp: new Date().toISOString()
          }
        },
        {
          headers: {
            'Authorization': userJWT, // CRITICAL: Use user's JWT, not RunPod key
            'Content-Type': 'application/json'
          },
          timeout: 30000 // 30 second timeout
        }
      );

      console.log('✅ TxAgent embedding response received');
      console.log(`📊 Response status: ${response.status}`);
      console.log(`📋 Response data keys: ${Object.keys(response.data)}`);

      // Handle different TxAgent response formats
      if (response.data.embedding && Array.isArray(response.data.embedding)) {
        console.log(`🎯 Direct embedding received: ${response.data.embedding.length} dimensions`);
        return response.data.embedding;
      }
      
      if (response.data.document_ids && response.data.document_ids.length > 0) {
        console.log(`📄 Background processing initiated: ${response.data.document_ids.length} documents`);
        // For background processing, we need to return a placeholder embedding
        // The actual embeddings will be stored in the database by the TxAgent
        throw new Error('TxAgent background processing - use fallback embedding');
      }
      
      throw new Error('Invalid TxAgent response format - no embedding found');
      
    } catch (error) {
      console.error('❌ TxAgent embedding failed:', error.message);
      
      if (error.code === 'ECONNABORTED') {
        throw new Error('TxAgent request timeout - container may be starting up');
      }
      
      if (error.response?.status === 401) {
        throw new Error('TxAgent authentication failed - invalid JWT token');
      }
      
      if (error.response?.status === 405) {
        throw new Error('TxAgent method not allowed - check container endpoints');
      }
      
      throw new Error(`TxAgent embedding failed: ${error.message}`);
    }
  }

  // LEGACY: Keep RunPod method for backward compatibility (but not used with JWT)
  async generateRunPodEmbedding(text) {
    try {
      console.log('⚠️  LEGACY: Using RunPod API key method (deprecated)');
      
      const response = await axios.post(
        this.runpodUrl,
        {
          input: {
            text: text,
            normalize: true,
            truncate: true
          }
        },
        {
          headers: {
            'Authorization': `Bearer ${this.runpodKey}`, // Legacy RunPod API key
            'Content-Type': 'application/json'
          },
          timeout: 30000 // 30 second timeout
        }
      );

      if (response.data && response.data.output && response.data.output.embedding) {
        return response.data.output.embedding;
      }
      
      throw new Error('Invalid RunPod response format');
      
    } catch (error) {
      if (error.code === 'ECONNABORTED') {
        throw new Error('RunPod request timeout');
      }
      throw new Error(`RunPod embedding failed: ${error.message}`);
    }
  }

  async generateOpenAIEmbedding(text) {
    try {
      console.log('🤖 Calling OpenAI embeddings API');
      console.log(`📝 Text length: ${text.length} characters`);
      
      const response = await axios.post(
        'https://api.openai.com/v1/embeddings',
        {
          model: 'text-embedding-3-small',
          input: text,
          encoding_format: 'float',
          dimensions: 768 // Force 768 dimensions to match Supabase documents table
        },
        {
          headers: {
            'Authorization': `Bearer ${this.openaiKey}`,
            'Content-Type': 'application/json'
          },
          timeout: 30000
        }
      );

      if (response.data && response.data.data && response.data.data[0]) {
        console.log(`✅ OpenAI embedding generated: ${response.data.data[0].embedding.length} dimensions`);
        console.log(`📊 Usage: ${JSON.stringify(response.data.usage)}`);
        return response.data.data[0].embedding;
      }
      
      throw new Error('Invalid OpenAI response format');
      
    } catch (error) {
      if (error.response) {
        throw new Error(`OpenAI API error: ${error.response.data?.error?.message || error.response.statusText}`);
      }
      throw new Error(`OpenAI embedding failed: ${error.message}`);
    }
  }

  // Utility method to calculate cosine similarity
  cosineSimilarity(vecA, vecB) {
    if (vecA.length !== vecB.length) {
      throw new Error('Vectors must have the same length');
    }

    let dotProduct = 0;
    let normA = 0;
    let normB = 0;

    for (let i = 0; i < vecA.length; i++) {
      dotProduct += vecA[i] * vecB[i];
      normA += vecA[i] * vecA[i];
      normB += vecB[i] * vecB[i];
    }

    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
  }
}